File: pypets/yage/capabilities/wall_crawler.py
"""
import SwiftUI

class WallCrawler: Capability {
    def install(on subject: Entity) {
        super().install(on: subject)
        subject.capability(for: BounceOnLateralCollisions.self)?.is_enabled = False
        subject.capability(for: FlipHorizontallyWhenGoingLeft.self)?.is_enabled = False
    }

    def do_update(self, collisions: List[Collision], time: float) {
        guard is_enabled else { return }
        guard let direction = subject?.direction else { return }

        let isGoingUp = direction.dy < -0.0001
        let isGoingRight = direction.dx > 0.0001
        let isGoingDown = direction.dy > 0.0001
        let isGoingLeft = direction.dx < -0.0001

        if isGoingUp and touchesScreenTop() {
            crawlAlongtop_bound()
            return
        }
        if isGoingRight and touchesScreenRight() {
            crawlUpright_bound()
            return
        }
        if isGoingDown and touchesScreenBottom() {
            crawlAlongbottom_bound()
            return
        }
        if isGoingLeft and touchesScreenLeft() {
            crawlDownleft_bound()
            return
        }
    }

    def _crawlAlongtop_bound() {
        guard let subject else { return }
        subject.direction = .init(dx: -1, dy: 0)
        subject.frame.origin = Point(x: subject.frame.origin.x, y: 0)
        subject.rotation?.isFlippedHorizontally = True
        subject.rotation?.isFlippedVertically = True
        subject.rotation?.z = 0
    }

    def _crawlUpright_bound() {
        guard let subject else { return }
        subject.direction = .init(dx: 0, dy: -1)
        subject.frame.origin = Point(
            x: subject.world_bounds.width - subject.frame.width,
            y: subject.frame.origin.y
        )
        subject.rotation?.isFlippedHorizontally = False
        subject.rotation?.isFlippedVertically = False
        subject.rotation?.z = .pi * 0.5
    }

    def _crawlAlongbottom_bound() {
        guard let subject else { return }
        subject.direction = Vector(1, 0)
        subject.frame.origin = Point(
            x: subject.frame.origin.x,
            y: subject.world_bounds.height - subject.frame.height
        )
        subject.rotation?.isFlippedHorizontally = False
        subject.rotation?.isFlippedVertically = False
        subject.rotation?.z = 0
    }

    def _crawlDownleft_bound() {
        guard let subject else { return }
        subject.direction = .init(dx: 0, dy: 1)
        subject.frame.origin = Point(x: 0, y: subject.frame.origin.y)
        subject.rotation?.isFlippedHorizontally = False
        subject.rotation?.isFlippedVertically = False
        subject.rotation?.z = .pi * 1.5
    }

    def _touchesScreenTop() -> bool {
        guard let subject else { return False }
        return subject.frame.min_y <= 0
    }

    def _touchesScreenRight() -> bool {
        guard let subject else { return False }
        let bound = subject.world_bounds.width
        return subject.frame.max_x >= bound
    }

    def _touchesScreenBottom() -> bool {
        guard let subject else { return False }
        let bound = subject.world_bounds.height
        return subject.frame.max_y >= bound
    }

    def _touchesScreenLeft() -> bool {
        guard let subject else { return False }
        return subject.frame.min_x <= 0
    }
}
"""

File: Packages/Yage/Sources/Yage/Capabilities/WallCrawler.swift
import SwiftUI

public class WallCrawler: Capability {
    override public func doUpdate(with collisions: Collisions, after time: TimeInterval) {
        guard case .move = subject?.state else { return }
        guard let direction = subject?.direction else { return }
        
        if isGoingUp(direction) && touchesScreenTop() {
            crawlAlongTopBound()
            return
        }
        if isGoingRight(direction) && touchesScreenRight() {
            crawlUpRightBound()
            return
        }
        if isGoingDown(direction) && touchesScreenBottom() {
            crawlAlongBottomBound()
            return
        }
        if isGoingLeft(direction) && touchesScreenLeft() {
            crawlDownLeftBound()
            return
        }
    }
    
    // MARK: - Direction
    
    private func isGoingUp(_ direction: CGVector) -> Bool {
        return direction.dy < -0.0001
    }
    
    private func isGoingRight(_ direction: CGVector) -> Bool {
        return direction.dx > 0.0001
    }
    
    private func isGoingDown(_ direction: CGVector) -> Bool {
        return direction.dy > 0.0001
    }
    
    private func isGoingLeft(_ direction: CGVector) -> Bool {
        return direction.dx < -0.0001
    }
    
    // MARK: - Collisions

    private func touchesScreenTop() -> Bool {
        guard let subject else { return false }
        return subject.frame.minY <= 0
    }

    private func touchesScreenRight() -> Bool {
        guard let subject else { return false }
        let bound = subject.worldBounds.width
        return subject.frame.maxX >= bound
    }

    private func touchesScreenBottom() -> Bool {
        guard let subject else { return false }
        let bound = subject.worldBounds.height
        return subject.frame.maxY >= bound
    }

    private func touchesScreenLeft() -> Bool {
        guard let subject else { return false }
        return subject.frame.minX <= 0
    }
    
    // MARK: - Crawling

    private func crawlAlongTopBound() {
        guard let subject else { return }
        subject.direction = .init(dx: -1, dy: 0)
        subject.frame.origin = CGPoint(x: subject.frame.origin.x, y: 0)
        subject.rotation?.isFlippedHorizontally = true
        subject.rotation?.isFlippedVertically = true
        subject.rotation?.zAngle = 0
    }

    private func crawlUpRightBound() {
        guard let subject else { return }
        subject.direction = .init(dx: 0, dy: -1)
        subject.frame.origin = CGPoint(
            x: subject.worldBounds.width - subject.frame.width,
            y: subject.frame.origin.y
        )
        subject.rotation?.isFlippedHorizontally = false
        subject.rotation?.isFlippedVertically = false
        subject.rotation?.zAngle = .pi * 0.5
    }

    private func crawlAlongBottomBound() {
        guard let subject else { return }
        subject.direction = .init(dx: 1, dy: 0)
        subject.frame.origin = CGPoint(
            x: subject.frame.origin.x,
            y: subject.worldBounds.height - subject.frame.height
        )
        subject.rotation?.isFlippedHorizontally = false
        subject.rotation?.isFlippedVertically = false
        subject.rotation?.zAngle = 0
    }

    private func crawlDownLeftBound() {
        guard let subject else { return }
        subject.direction = .init(dx: 0, dy: 1)
        subject.frame.origin = CGPoint(x: 0, y: subject.frame.origin.y)
        subject.rotation?.isFlippedHorizontally = false
        subject.rotation?.isFlippedVertically = false
        subject.rotation?.zAngle = .pi * 1.5
    }
}
